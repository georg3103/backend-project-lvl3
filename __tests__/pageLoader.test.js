import {
  test, expect, describe, beforeEach,
} from '@jest/globals';
import { promises as fsPromises } from 'fs';
import os from 'os';
import path from 'path';
import nock from 'nock';
import downloadPage from '../src';

const link = 'https://test.com';
const resourcesFolder = 'test-com_files';
let pathToTempDir;
let pathToResources;

const getFixturePath = (fileName) => path.join('__tests__/__fixtures__', fileName);

const readFile = async (pathToFile) => {
  const file = await fsPromises.readFile(pathToFile, { encoding: 'utf8' });
  return file;
};

beforeAll(() => nock.disableNetConnect());

beforeEach(async () => {
  pathToTempDir = await fsPromises.mkdtemp(
    path.join(os.tmpdir(), 'page-loader-'),
  );
  pathToResources = `${pathToTempDir}/${resourcesFolder}`;
});

describe('pageLoader functionality', () => {
  test('link with subpath nested resources has been downloaded', async () => {
    const html = await readFile(getFixturePath('index.html'));
    const expectedHtml = await readFile(getFixturePath('changedIndex.html'));
    const expectedStyle = 'a { color: red }';
    const expectedScript = 'console.log("Hello world!")';

    nock(`${link}/tests/`)
      .get('/')
      .reply(200, html)
      .get('/style.css')
      .reply(200, expectedStyle)
      .get('/folder/script.txt')
      .reply(200, expectedScript);

    await downloadPage(`${link}/tests/`, pathToTempDir);

    const changedHtml = await readFile(`${pathToTempDir}/test-com.html`);
    const loadedStyle = await readFile(`${pathToResources}/style.css`);
    const loadedScript = await readFile(`${pathToResources}/folder-script.txt`);

    expect(changedHtml).toBe(expectedHtml);
    expect(loadedStyle).toBe(expectedStyle);
    expect(loadedScript).toBe(expectedScript);
  });
});

describe('pageLoader error handling', () => {
  test('passed link to non-existent site', async () => {
    const notValidLink = 'http://qwertyui12345.com/';

    nock(notValidLink).get('/').reply(404);

    await expect(downloadPage(notValidLink, pathToTempDir)).rejects.toThrowErrorMatchingSnapshot();
  });
});
