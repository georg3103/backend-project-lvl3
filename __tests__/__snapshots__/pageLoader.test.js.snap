// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pageLoader error handling passed link to non-existent site 1`] = `"Request failed with status code 404"`;
